### PR TITLE
Fix missing events in OnRecvPacket

### PR DESCRIPTION
**Pull request by bluele**
## Description

The events set in the cache context need to be set in the original context.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/core/keeper/msg_server.go [hunk 1]
```diff
@@ -503,6 +503,7 @@ func (k Keeper) RecvPacket(goCtx context.Context, msg *channeltypes.MsgRecvPacke
 	if ack == nil || ack.Success() {
 		// write application state changes for asynchronous and successful acknowledgements
 		writeFn()
+		ctx.EventManager().EmitEvents(cacheCtx.EventManager().Events())
```
Comment (colin-axner) [inline modules/core/keeper/msg_server.go:506]: could you add a comment that indicates that `ctx.CacheContext()` sets a new event manager and thus won't cause duplicate events to be returned by `cacheCtx.EventManager().Events()`

At first I thought this line might cause events emitted by `k.ChannelKeeper.RecvPacket` to be represented twice but I realized the cache context reset the events by setting a new event manager

It'd be great to have this written in the code

**Review by colin-axner — APPROVED** [review]
Very nice catch!! Just left a suggestion for adding a in code comment for easier verifiability of the code. Tested ACK

**Review by colin-axner — CHANGES_REQUESTED** [review]
On second thought, I think we should emit events in all cases. So we should move the line added above the if statement. This allows failed acks to be emitted as well

**Comment by AdityaSripal** [comment]
Great catch! Agreed with Colin that we should emit events regardless of callback success. We must document to developers that state gets reverted but any emitted events still get emitted even in error case

**Comment by bluele** [comment]
@colin-axner
Thanks for your comments! I've fixed them and squashed commits.

File: modules/core/keeper/msg_server.go [hunk 2]
```diff
@@ -500,6 +500,9 @@ func (k Keeper) RecvPacket(goCtx context.Context, msg *channeltypes.MsgRecvPacke
 	// Cache context so that we may discard state changes from callback if the acknowledgement is unsuccessful.
 	cacheCtx, writeFn := ctx.CacheContext()
 	ack := cbs.OnRecvPacket(cacheCtx, msg.Packet, relayer)
+	// This doesn't cause duplicate events to be emitted.
+	// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
```
Comment (AdityaSripal) [inline modules/core/keeper/msg_server.go:504]:
```suggestion
	// NOTE: The context returned by CacheContext() refers to a new EventManager, so it needs to explicitly set events to the original context.
	// Events from callback are emitted regardless of acknowledgement success
```

**Review by AdityaSripal — APPROVED** [review]

**Review by colin-axner — APPROVED** [review]
LGTM, thanks for the contribution @bluele and the fast response time!

**Pull request merged**